File: library/src/main/java/com/jude/spinnerlibrary/ASpinner.java
package com.jude.spinnerlibrary;

import android.content.Context;
import android.util.AttributeSet;
import android.widget.Spinner;

public class ASpinner extends Spinner
{
    private OnItemSelectedListener listener;
    private OnSpinnerEventsListener mListener;
    private boolean mOpenInitiated = false;

    public ASpinner(Context context)
    {
        super(context);
    }

    public ASpinner(Context context, int mode)
    {
        super(context, mode);
    }

    public ASpinner(Context context, AttributeSet attrs)
    {
        super(context, attrs);
    }

    @Override
    public void setSelection(int position, boolean animate)
    {
        boolean sameSelected = position == getSelectedItemPosition();
        super.setSelection(position, animate);
        if (sameSelected)
        {
            if (listener != null)
            {
                listener.onItemSelected(this, getSelectedView(), position, getSelectedItemId());
            }
        }
    }

    @Override
    public void setSelection(int position)
    {
        super.setSelection(position);
        if (listener != null)
        {
            listener.onItemSelected(this, getSelectedView(), position, getSelectedItemId());
        }
    }

    public void setOnItemSelectedEvenIfUnchangedListener(OnItemSelectedListener listener)
    {
        this.listener = listener;
    }

    // =======================================================

    public interface OnSpinnerEventsListener
    {
        void onSpinnerOpened(Spinner spin);
        void onSpinnerClosed(Spinner spin);
    }

    @Override
    public boolean performClick()
    {
        mOpenInitiated = true;
        if (mListener != null)
        {
            mListener.onSpinnerOpened(this);
        }
        return super.performClick();
    }

    public void setSpinnerEventsListener(OnSpinnerEventsListener onSpinnerEventsListener)
    {
        mListener = onSpinnerEventsListener;
    }

    public void performClosedEvent()
    {
        mOpenInitiated = false;
        if (mListener != null)
        {
            mListener.onSpinnerClosed(this);
        }
    }

    public boolean hasBeenOpened()
    {
        return mOpenInitiated;
    }

    @Override
    public void onWindowFocusChanged(boolean hasWindowFocus)
    {
        super.onWindowFocusChanged(hasWindowFocus);
        if (hasBeenOpened() && hasWindowFocus)
        {
            performClosedEvent();
        }
    }
}

File: library/src/main/java/com/jude/spinnerlibrary/ASpinner.kt
package com.jude.spinnerlibrary

import android.content.Context
import android.util.AttributeSet
import android.widget.Spinner



class ASpinner : Spinner {
    private var listener: OnItemSelectedListener? = null
    private var mListener: SpinnerEventsListener? = null
    private var mOpenInitiated = false

    constructor(context: Context) : super(context) {}

    constructor(context: Context, mode: Int): super(context, mode) {}

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs) {}

    override fun setSelection(position: Int, animate: Boolean) {
        val sameSelected = position == selectedItemPosition
        super.setSelection(position, animate)
        if (sameSelected) {
            listener?.onItemSelected(this, selectedView, position, selectedItemId)
        }
    }

    override fun setSelection(position: Int) {
        super.setSelection(position)
        listener?.onItemSelected(this, selectedView, position, selectedItemId)
    }

    fun setOnItemSelectedEvenIfUnchangedListener(listener: OnItemSelectedListener) {
        this.listener = listener
    }

    override fun performClick(): Boolean {
        mOpenInitiated = true
        mListener?.onSpinnerOpened(this)
        return super.performClick()
    }

    fun setSpinnerEventsListener(onSpinnerEventsListener: SpinnerEventsListener) {
        mListener = onSpinnerEventsListener
    }

    private fun performClosedEvent() {
        mOpenInitiated = false
        mListener?.onSpinnerClosed(this)
    }

    private fun hasBeenOpened(): Boolean {
        return mOpenInitiated
    }

    override fun onWindowFocusChanged(hasWindowFocus: Boolean) {
        super.onWindowFocusChanged(hasWindowFocus)
        if (hasBeenOpened() && hasWindowFocus) {
            performClosedEvent()
        }
    }
}
